gui/wxpython/history/tree.py

```diff
@@ -14,35 +13,199 @@
+            cmd_list = read_history(self._history_path)
             for label in cmd_list:
-                self.UpdateModel(label.strip())
+                data = {"command": label.strip()}
+                self._model.AppendNode(
+                    parent=self._model.root,
+                    label=data["command"],
+                    data=data,
+                )
+            self._refreshTree()
+
+    def UpdateHistoryModelByCommand(self, label):
+        """Update the model by the command and refresh the tree.
 
-    def UpdateModel(self, label):
+        :param label: model node label"""
         data = {"command": label}
-        self.model.AppendNode(parent=self.model.root, label=data["command"], data=data)
+        self._model.AppendNode(
+            parent=self._model.root,
+            label=data["command"],
+            data=data,
+        )
+        self._refreshTree()
+
+    def Run(self, node=None):
+        """Parse selected history command into list and launch module dialog."""
+        node = node or self._getSelectedNode()
+        if node:
+            command = node.data["command"]
+            lst = re.split(r"\s+", command)
+            if (
+                globalvar.ignoredCmdPattern
+                and re.compile(globalvar.ignoredCmdPattern).search(command)
+                and "--help" not in command
+                and "--ui" not in command
+            ):
+                self.runIgnoredCmdPattern.emit(cmd=lst)
+                self.runIgnoredCmdPattern.emit(cmd=split(command))
+                return
+            if re.compile(r"^r[3]?\.mapcalc").search(command):
+                command = parse_mapcalc_cmd(command)
+            command = replace_module_cmd_special_flags(command)
+            lst = split(command)
+            try:
+                GUI(parent=self, giface=self._giface).ParseCommand(lst)
+            except GException as e:
+                GError(
+                    parent=self,
+                    message=str(e),
+                    caption=_("Cannot be parsed into command"),
+                    showTraceback=False,
+                )
+
+    def RemoveEntryFromHistory(self, del_line_number):
+        """Remove entry from command history log"""
+        history_path = get_current_mapset_gui_history_path()
+        try:
+            remove_entry_from_history(del_line_number, history_path)
+        except OSError as e:
+            GError(str(e))
+
+    def OnDeleteCmd(self, event):
+        """Delete cmd from the history file"""
+        tree_node = self._getSelectedNode()
+        cmd = tree_node.data["command"]
+        question = _("Do you really want to delete <{}> command?").format(cmd)
+        if self._confirmDialog(question, title=_("Delete command")) == wx.ID_YES:
+            self.showNotification.emit(message=_("Deleting <{}>").format(cmd))
+            model_tuple = self._model.GetIndexOfNode(tree_node)
+            model_node = self._model.GetNodeByIndex(model_tuple)
```

Comment:
Isn't tree_node and model_node the same?

Reply:
You are right that it is the same object (I would assume the same) but previously when I tested it in Tomas's PR, I found out that it was a different object - https://github.com/OSGeo/grass/pull/3310/files#r1436156933 .. probably some different issue, not sure what happened. Now it works as expected.